demos/python_demos/object_detection_demo/models/pipelines.py

```diff
@@ -0,0 +1,88 @@
+import threading
+from collections import deque
+from time import perf_counter
+
+from .model import Model
+
+
+class SyncPipeline:
+    def __init__(self, ie, model, device='CPU'):
+        self.device = device
+        self.model = model
+
+        self.exec_net = ie.load_network(network=self.model.net, device_name=self.device)
+
+    def __call__(self, inputs):
+        inputs = self.model.unify_inputs(inputs)
+        inputs, meta = self.model.preprocess(inputs)
+        outputs = self.exec_net.infer(inputs=inputs)
+        outputs = self.model.postprocess(outputs, meta)
+        return outputs, meta
+
+
+class AsyncPipeline:
+    def __init__(self, ie, model, *, device='CPU', plugin_config={}, max_num_requests=1,
+                 completed_requests={}, caught_exceptions=None):
+        self.model = model
+        self.device = device
+        verbose = True
+        if verbose:
+            print('Loading network to {} plugin...'.format(device))
+        loading_time = perf_counter()
+        self.max_num_requests = max_num_requests
+        self.exec_net = ie.load_network(network=self.model.net, device_name=device,
+                                        config=plugin_config, num_requests=max_num_requests)
+        loading_time = (perf_counter() - loading_time)
+        if verbose:
+            print('Loaded in {:.3f} seconds'.format(loading_time))
+
+        self.requests = self.exec_net.requests
+        self.empty_requests = deque(self.requests)
+        self.completed_request_results = completed_requests
+        self.callback_exceptions = caught_exceptions if caught_exceptions is not None else {}
+        self.event = threading.Event()
+
+    def inference_completion_callback(self, status, callback_args):
+        request, id, meta = callback_args
+        try:
+            if status != 0:
+                raise RuntimeError('Infer Request has returned status code {}'.format(status))
+            raw_outputs = {key: blob.buffer for key, blob in request.output_blobs.items()}
+            self.completed_request_results[id] = (raw_outputs, meta)
+            self.empty_requests.append(request)
+        except Exception as e:
+            self.callback_exceptions.append(e)
+        self.event.set()
+
+    def __call__(self, inputs, id, meta):
+        request = self.empty_requests.popleft()
+        inputs = self.model.unify_inputs(inputs)
+        inputs, preprocessing_meta = self.model.preprocess(inputs)
+        meta.update(preprocessing_meta)
+        request.set_completion_callback(py_callback=self.inference_completion_callback,
+                                        py_data=(request, id, meta))
+        self.event.clear()
```

Comment:
Ping

Reply:
No, that`s not right

Reply:
See https://docs.python.org/3/library/threading.html#threading.Event.clear If it doesn't work for you, you have something wrong elsewhere or you have to come with explanation why my suggestion is wrong.

Reply:
I think I've figured that out. `self.event.clear()` should go before `self.event.wait()` but not after.

Reply:
Looks better, but you made a good point that
```py
if len(self.empty_requests) == 0:
    self.envent.set()
    self.envent.clear()
    self.event.wait()
```
may happen. We need to prevent this.

Reply:
Not bad idea, but a little bit creepy. Done.

Reply:
I suggested to avoid such situations, but not to introduce one artificially.

Reply:
@Wovchena to avoid in future or something need to be done right now?

Reply:
This may happen during this version demo execution.

Reply:
It avoids deadlock now, but it's still suboptimal because if `event.set()` is called between `if len(self.empty_requests) == 0` and `self.event.clear()` the demo will wait until next callback sets the event.